### PR TITLE
Implement sticky headers using CSS

### PR DESCRIPTION
**Pull request by gsouquet**
Fixes vector-im/element-web#14709
Fixes https://github.com/vector-im/element-web/issues/14665
Fixes https://github.com/vector-im/element-web/issues/14531
Fixes https://github.com/vector-im/element-web/issues/14493
Fixes https://github.com/vector-im/element-web/issues/17412
Fixes https://github.com/vector-im/element-web/issues/17283

Asking for a design review by @matrix-org/design as implementing this requires us to remove the blurred background image as we need to have a background color set on the sublist header for this CSS implementation to work.
As an alternative to a solid background we could have a horizontal linear gradient.

Tested on
- Chromium
- Firefox
- Safari

Overall the scrolling experience is way smoother. Looking at the recording it can seem that we are doing more work with the JavaScript implementation, but this is purely because now the browser is able to render at a higher frame rate

# Before 🐌 

Screen Shot 2021-05-18 at 10 35 58

# After 🐎 

Screen Shot 2021-05-18 at 10 38 36

---
This PR currently has no changelog labels, so will not be included in changelogs.

Add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is plus `X-Breaking-Change` if it's a breaking change.

### CODE REVIEW

**Comment by SimonBrandner** [comment]
Also fixes [#14493](https://github.com/vector-im/element-web/issues/14493)

File: res/themes/light/css/_mods.scss [hunk 1]
```diff
@@ -1,30 +1,3 @@
-// sidebar blurred avatar background
-//
-// if backdrop-filter is supported,
-// set the user avatar (if any) as a background so
-// it can be blurred by the tag panel and room list
-
-@supports (backdrop-filter: none) {
-    .mx_LeftPanel {
-        background-image: var(--avatar-url, unset);
```
Comment (t3chguy) [inline res/themes/light/css/_mods.scss:9]: Can get rid of where this var is set too https://github.com/matrix-org/matrix-react-sdk/pull/6054/files#diff-bb5eeeda9a8d5f7b2a4cbcce491cfbf482b2c704207b28a167d31992851e7c1aL128-L143

File: res/css/views/rooms/_RoomSublist.scss [hunk 2]
```diff
@@ -27,52 +27,16 @@ limitations under the License.
...
-        // all works by ensuring the header text has a fixed height when sticky so the
-        // fixed height of the container can maintain the scroll position.
-
-        // The combined height must be set in the LeftPanel component for sticky headers
-        // to work correctly.
-        padding-bottom: 8px;
+        position: sticky;
+        top: 0;
+        z-index: 9;
+        background: rgb(245,245,245);
```
Comment (t3chguy) [inline res/css/views/rooms/_RoomSublist.scss:33]: this ought to be a theme var

File: res/css/views/rooms/_RoomSublist.scss [hunk 3]
```diff
@@ -27,52 +27,16 @@ limitations under the License.
...
-        // target the headerText instead of the container to reduce jumps when scrolling,
-        // and to help hide the badges/other buttons that could appear on hover. This
-        // all works by ensuring the header text has a fixed height when sticky so the
-        // fixed height of the container can maintain the scroll position.
-
-        // The combined height must be set in the LeftPanel component for sticky headers
-        // to work correctly.
-        padding-bottom: 8px;
+        position: sticky;
+        top: 0;
```
Comment (t3chguy) [inline res/css/views/rooms/_RoomSublist.scss:31]: does this sticky both top and bottom like existing impl?

image

**Comment by SimonBrandner** [comment]
Screenshot_20210518_122255

The headers seem to be missing the correct color for me...

**Comment by t3chguy** [comment]
@SimonBrandner that looks like https://github.com/matrix-org/matrix-react-sdk/pull/6054#discussion_r634237439

**Comment by SimonBrandner** [comment]
Ah, missed it

**Comment by nadonomy** [comment]
@gsouquet since we spoke last via DM on this, did you end up finding a solution for the bottom/next section headers?

**Comment by gsouquet** [comment]
@nadonomy A full CSS solution will not be possible as the property make elements stick at the top of the visible area of the elements. However we can explore more performant ways of approaching this using an `IntersectionObserver`
This has been deprioritized as we were not really sure on the way forward and the implications for design

**Comment by nadonomy** [comment]
> @nadonomy A full CSS solution will not be possible as the property make elements stick at the top of the visible area of the elements. However we can explore more performant ways of approaching this using an `IntersectionObserver` This has been deprioritized as we were not really sure on the way forward and the implications for design

@gsouquet apologies for not following up explicitly on this. When developing this version of the room list we found the bottom sticky headers were essential. Keen to not rip them out when refactoring. Thoughts:

1. Taking the above as a requirement, if you're able to reprioritise the `IntersectionObserver` approach, it'd be great for you to pair with @janogarcia on whatever issues you find, and I can make time to aid in support/review.
2. Room list navigation is fundamentally changing having shipped Spaces. We're expecting users to need to navigate smaller lists. We're now embarking on a project to improve the overall info architecture of the apps, which includes the room list. It's possible we might no longer need the labels, but we won't know after ideation and testing.

If the effort for 1 is low, it could be worth finish lining this PR with @janogarcia. If you'd prefer to see where arrive at with the info architecture work this quarter, we could liaise between yourself, @t3chguy & @niquewoodhouse.

What are your thoughts?

**Pull request closed without merging**